## [특강] Numpy, Pandas, Matplotlib

### 1. Numpy: 고속 연산, 슬라이싱 및 차원 제어
#### 1.1 명시적 타입 배열 생성과 다차원 슬라이싱

배열(Array)
- 같은 자료형 (ex. int64 = 8byte, float64 = 8byte) 의 데이터가 연속된 메모리 공간에 저장됨  ==> 1차원 배열, '선'형 구조
- 각 원소의 크기(방의 크기)가 동일함 -> 인덱스->메모리 주소 계산이 가능 = 임의의 위치에 빠르게 접근 ==> 같은 자료형만 저장하는 이유?

    - index 0 -> 8 bit만큼 이동 -> index 1 -> 8 bit 이동 -> index 2 ....
- 인덱스 연산자([]) 접근 가능
- numpy: C언어로 개발(직접 메모리 접근)
- dtype: 자료형 명시 ( ex> dtype=np.int64 또는 np.float64)

* 리스트
- 리스트 자체에는 객체를 가르키는 참조가 지정됨
- 실제 객체는 메모리에 다른 위치에 존재 가능
- 리스트 내부에는 참조들이 연속적으로 저장됨


방의 크기
- 비트(bit): 0.1
- 바이트(byte): 비트x8 ex) 1byte = 8bit.


부호비트(sign bit): 숫자가 양수/음수 판단하는 비트. 8비트에서 가장 왼쪽 비트 ex) 0 = 양수, 1 = 음수


numpy
- 문자열 x, 숫자만 입력 가능 (array는 같은 자료형의 데이터만 저장 ==> 1차원 배열 '선'형 구조)


    * 2차원 배열(행, 열) 존재, 즉 면적 존재. `matrix`
        - [0번 행][0번 열]부터 시작 
        - ex) [0][0], [1][0] 1행 0열
    * 1차원 배열 - 선 `vector`

    * 3차원 이상의 배열 `Tensor`



In [ ]:
import numpy as np # 보통 numpy = np로 칭함

nums = np.array([10, 20, 30], dtype=np.float64) # 1차원 배열(Vector), dtype으로 자료형 결정

nums

In [ ]:
nums[0]

In [ ]:
#행렬 = matrix 2차원 배열

matrix = np.array([
    [10, 20, 30],
    [40, 50, 60],
    [70, 80, 90]
])

#### 배열 선택, 슬라이싱
- `matrix[행, 열]`
- `matrix[0:2, 1:3]`: 범위 지정 가능 ex) 0행~1행까지, 1행~2열까지
- `matrix[:, 0]`: 모든 행의 0번째 열을 추출
- `matrix[1:, :2]`: 1번째 행부터 끝까지, 0번째 열부터 1번째 열까지 

In [ ]:
print(matrix[0, 2]) # 2차원 배열 0행 2열 출력, 대괄호 1개 또는 2개로 사용 가능
print(matrix[0][2])

print(matrix[0:2, 1:3])
print(matrix[:, 0])
print(matrix[1:, :2])


#### 1.2 차원 변경(reshape, flatten)과 축(Axis)

reshape(행, 열) / flatten()

- `reshape(행, 열)` - 1차원 배열 -> 지정된 행과 열 구조로 차원수를 변경
    - `-1`: 자동 결정
- `arrange(시작, 종료, 증감단위)`: 시작 이상 종료 미만으로 숫자를 생성



- `flatten()`: 2차원 배열 -> 1차원으로 평탄화

In [ ]:
arr = np.arange(1, 13)
print(arr)
mat_3x4 = arr.reshape(3, -1)
print(mat_3x4)
mat_3x4_2 = arr.reshape(-1, 4)
print(mat_3x4_2)

flat_mat = mat_3x4.flatten()
print(flat_mat)

축(axis)

2차원 배열에서 연산의 방향을 설정
- axis: 0 -> 행 기준에서 연산
- axis: 1 -> 열 기준에서 연산

In [ ]:
mat = np.arange(1, 7).reshape(2, -1) # 1차원 arange, reshape로 matrix 만들기
print(mat)

sum_axis0 = mat.sum(axis=0) # 행 기준 합계

sum_axis1 = mat.sum(axis=1) # 열 기준 합계

print("행 기준 합계:", sum_axis0)
print("열 기준 합계:", sum_axis1)

#### 1.3 Numpy 브로드캐스팅(Broadcasting) 연산

In [ ]:
mat = np.array([
    [10, 20, 30],
    [40, 50, 60]
])

print(mat * 5) # 전체 데이터에 동일한 값 연산 = 스칼라 브로드캐스팅
print(mat + 5)


In [ ]:
vec = np.array([1, 2, 3])
print(mat + vec) # matrix + vector = matrix의 모든 데이터에 vector값을 더하기


#### 1.4 행렬 전치(Transpose) 및 행렬곱(@)

- `객체.transpose()` -> {행, 열} -> (열, 행). **전치**
    - = `객체.T`

In [ ]:
mat = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

print(mat)
print("전치 행렬:\n", mat.transpose())
print("전치 행렬:\n", mat.T)

In [ ]:
# 행렬곱 연산 
mat1 = np.array([[10, 20], [30, 40]])
mat2 = np.array([[1, 2], [3, 4]])

mat3 = mat1 @ mat2
print(mat3)

[[ 70 100]
 [150 220]]


$$
\begin{bmatrix}
10 & 20 \\
30 & 40
\end{bmatrix}
\begin{bmatrix}
1 & 2 \\
3 & 4
\end{bmatrix}
=
\begin{bmatrix}
10\times1 + 20\times3 & 10\times2 + 20\times4 \\
30\times1 + 40\times3 & 30\times2 + 40\times4
\end{bmatrix}
=
\begin{bmatrix}
70 & 100 \\
150 & 220
\end{bmatrix}
$$